doc/developer/design/20220303_secrets.md

```diff
@@ -0,0 +1,214 @@
+    ///
+    /// Implementations must apply the operations atomically. If the method
+    /// returns `Ok(())`, then all operations have been applied successfully;
+    /// if the method returns `Err(())`, then none of the operations have been
+    /// applied.
+    ///
+    /// Implementations are permitted to reject combinations of operations which
+    /// they cannot apply atomically.
+    fn apply(&mut self, ops: Vec<SecretOp>) -> Result<(), anyhow::Error>;
+}
+
+/// An operation on a [`SecretsController`].
+pub enum SecretOp {
+    /// Create or update the contents of a secret.
+    Ensure {
+        /// The ID of the secret to create or update.
+        id: GlobalId,
+        /// The binary contents of the secret.
+        contents: Vec<u8>,
+    },
+    /// Delete a secret.
+    Delete {
+        /// The id of the secret to delete.
+        id: GlobalId
+    }
+}
+```
+
+There will be two concrete implementations of the `SecretsController`:
+
+  * The `FilesystemSecretsController` will store secrets on the local filesystem.
+    It will not support multi-operation atomicity, but will support atomically
+    updating or removing one secret at a time.
+
+  * The `KubernetesSecretsController` will store secrets in Kubernetes. It
+    will support multi-operation atomicity.
+
+All secrets controllers expose the contents of secrets to the [`STORAGE`]
+and [`COMPUTE`] layer by way of a directory that contains one file per secret,
+where the name of the file is the secret's ID and the contents of the file are
+the secret's contents. For example, if IDs `u12` and `u36` are secrets, the
+directory would look like this:
+
+```
+secrets/
+  u12
+  u36
+```
+
+This bears repeating: even the `KubernetesSecretsController` will expose the
+contents of the secrets via the filesystem. The filesystem is the *lingua
+franca* for exposing secrets to applications. Kubernetes supports exposing the
+contents of secrets as files on the filesystem, as does [HashiCorp Vault].
+
+The [`STORAGE`] and [`COMPUTE`] layers will need to be taught to re-read the
+contents of secrets that they consume from disk periodically, in case the
+contents have bene updated by an `ALTER SECRET`. We may wish to build a helper
+"secret consumer" library that uses filesystem events to notice when the secrets
+change, rather than polling periodically.
+
+### Secrets configuration
+
+The [`ADAPTER`] layer must be configured with a `SecretsController`
+implementation. The [`STORAGE`] and [`COMPUTE`] layers must be configured
+with the path on the local filesystem in which the secrets controller has made
+the contents of the secrets available.
+
+That means that presently `materialized` will create a
+`FilesystemSecretsController` that stores data in `mzdata/secrets`. For
+simplicity, to start, this directory will not be configurable. This controller
+will be provided to the `Coordinator`, while the path to `mzdata/secrets` will
+be provided to the dataflow layer.
```

Comment:
what does the dataflow layer use secrets for? Currently, I only know of the Kafka source that requires secrets.

Reply:
The Kafka sink also needs access to secrets. It is possible that in the long term STORAGE will need access to secrets while COMPUTE will not, but for now sinks live in COMPUTE rather than STORAGE.